doc/man/man8/slappasswd.8
.TH SLAPPASSWD 8C "RELEASEDATE" "OpenLDAP LDVERSION"
.\" Copyright 1998-2026 The OpenLDAP Foundation All Rights Reserved.
.\" Copying restrictions apply.  See COPYRIGHT/LICENSE.
.\" $OpenLDAP$
.SH NAME
slappasswd \- OpenLDAP password utility
.SH SYNOPSIS
.B SBINDIR/slappasswd
[\c
.BR \-v ]
[\c
.BR \-u ]
[\c
.BR \-g \||\| \-s \ \fIsecret\fR \||\| \fB\-T \ \fIfile\fR]
[\c
.BI \-h \ hash\fR]
[\c
.BI \-c \ salt-format\fR]
[\c
.BR \-n ]
[\c
.BI \-o \ option\fR[ = value\fR]]
.LP
.SH DESCRIPTION
.LP
.B Slappasswd
is used to generate an userPassword value
suitable for use with
.BR ldapmodify (1),
.BR slapd.conf (5)
.I rootpw
configuration directive or the 
.BR slapd\-config (5) 
.I olcRootPW
configuration directive.
.
.SH OPTIONS
.TP
.B \-v
enable verbose mode.
.TP
.B \-u
Generate RFC 2307 userPassword values (the default).  Future
versions of this program may generate alternative syntaxes
by default.  This option is provided for forward compatibility.
.TP
.BI \-s \ secret
The secret to hash.
If this,
.B \-g
and
.B \-T
are absent, the user will be prompted for the secret to hash.
.BR \-s ,
.B \-g
and
.B \-T
are mutually exclusive flags.
.TP
.BI \-g
Generate the secret.
If this,
.B \-s
and
.B \-T
are absent, the user will be prompted for the secret to hash.
.BR \-s ,
.B \-g
and
.B \-T
are mutually exclusive flags.
If this is present,
.I {CLEARTEXT}
is used as scheme.
.B \-g
and
.B \-h
are mutually exclusive flags.
.TP
.BI \-T \ "file"
Hash the contents of the file.
If this,
.B \-g
and
.B \-s
are absent, the user will be prompted for the secret to hash.
.BR \-s ,
.B \-g
and
.B \-T
and mutually exclusive flags.
.TP
.BI \-h \ "scheme"
If \fB\-h\fP is specified, one of the following RFC 2307 schemes may
be specified:
.BR {CRYPT} ,
.BR {MD5} ,
.BR {SMD5} ,
.BR {SSHA} ", and"
.BR {SHA} .
The default is 
.BR {SSHA} .

Note that scheme names may need to be protected, due to
.B {
and
.BR } ,
from expansion by the user's command interpreter.

.B {SHA}
and
.B {SSHA}
use the SHA-1 algorithm (FIPS 160-1), the latter with a seed.

.B {MD5}
and
.B {SMD5}
use the MD5 algorithm (RFC 1321), the latter with a seed.

.B {CRYPT}
uses the
.BR crypt (3).

.B {CLEARTEXT}
indicates that the new password should be added to userPassword as
clear text.
Unless
.I {CLEARTEXT}
is used, this flag is incompatible with option
.BR \-g .
.TP
.BI \-c \ crypt-salt-format
Specify the format of the salt passed to
.BR crypt (3)
when generating {CRYPT} passwords.  
This string needs to be in
.BR sprintf (3)
format and may include one (and only one)
.B %s
conversion.
This conversion will be substituted with a string of random
characters from [A\-Za\-z0\-9./].  For example,
.RB ' %.2s '
provides a two character salt and
.RB ' $1$%.8s '
tells some
versions of
.BR crypt (3)
to use an MD5 algorithm and provides
8 random characters of salt.
The default is
.RB ' %s ' ,
which provides 31 characters of salt.
.TP
.BI \-n
Omit the trailing newline; useful to pipe the credentials
into a command.
.TP
.BI \-o \ option\fR[ = value\fR]
Specify an
.I option
with a(n optional)
.IR value .
Possible generic options/values are:
.LP
.nf
              module\-path=<pathspec> (see `\fBmodulepath\fP' in slapd.conf(5))
              module\-load="<filename> [<arguments>...]" (see `\fBmoduleload\fP' in slapd.conf(5))

.in
You can load a dynamically loadable password hash module by
using this option.
.SH LIMITATIONS
The practice of storing hashed passwords in userPassword violates
Standard Track (RFC 4519) schema specifications and may hinder
interoperability.  A new attribute type, authPassword, to hold
hashed passwords has been defined (RFC 3112), but is not yet
implemented in
.BR slapd (8).
.LP
It should also be noted that the behavior of
.BR crypt (3)
is platform specific.
.SH "SECURITY CONSIDERATIONS"
Use of hashed passwords does not protect passwords during
protocol transfer.  TLS or other eavesdropping protections
should be in-place before using LDAP simple bind.
.LP
The hashed password values should be protected as if they
were clear text passwords.
.SH "SEE ALSO"
.BR ldappasswd (1),
.BR ldapmodify (1),
.BR slapd (8),
.BR slapd.conf (5),
.BR slapd\-config (5),
.B RFC 2307\fP,
.B RFC 4519\fP,
.B RFC 3112
.LP
"OpenLDAP Administrator's Guide" (http://www.OpenLDAP.org/doc/admin/)
.SH ACKNOWLEDGEMENTS
.so ../Project


CHANGES
OpenLDAP 2.6 Change Log

OpenLDAP 2.6.13 Engineering

OpenLDAP 2.6.12 Release (2026/01/29)
	Fixed libldap to reject empty types in LDIF (ITS#10429)
	Fixed libldap to not scroll past nul bytes (ITS#10430)
	Fixed libldap to enforce stop when encountering nul-leading line (ITS#10431)

OpenLDAP 2.6.11 Release (2026/01/28)
	Fixed slapd to use fresh timestamp for lastbind (ITS#10379)
	Fixed slapd delta-syncrepl to always use logDB rootdn (ITS#10360)
	Fixed slapd reverse lookup of proxied IPv6 addresses (ITS#10387)
	Fixed slapd logging buffer overflow (ITS#10410)
	Fixed slapd-ldap response when invalid secprops is configured (ITS#10392)
	Fixed slapd-mdb error when deleting last child of a branch (ITS#10304)
	Fixed slapd-mdb check for pool pause in search (ITS#10191)
	Fixed slapo-memberof clash with refint on subtree rename (ITS#10398)
	Fixed slapo-syncprov use correct rootDN for accesslog replay (ITS#10385)
	Minor Cleanup
		ITS#7901
		ITS#10329
		ITS#10335
		ITS#10339
		ITS#10343
		ITS#10344
		ITS#10345
		ITS#10347
		ITS#10348
		ITS#10349
		ITS#10353
		ITS#10358
		ITS#10359
		ITS#10366
		ITS#10367
		ITS#10369
		ITS#10370
		ITS#10371
		ITS#10372
		ITS#10374
		ITS#10375
		ITS#10376
		ITS#10377
		ITS#10379
		ITS#10380
		ITS#10381
		ITS#10384
		ITS#10388
		ITS#10390
		ITS#10391
		ITS#10400
		ITS#10401
		ITS#10408

OpenLDAP 2.6.10 Release (2025/05/22)
	Added slapd microsecond timestamp format for local logging (ITS#10140)
	Fixed libldap ldap_result behavior with LDAP_MSG_RECEIVED (ITS#10229)
	Fixed lloadd handling of starttls critical (ITS#10323)
	Fixed slapd syncrepl when used with slapo-rwm (ITS#10290)
	Fixed slapd regression with certain searches (ITS#10307)
	Fixed slapo-autoca olcAutoCAserverClass object (ITS#10288)
	Fixed slapo-pcache caching behaviors (ITS#10270)
	Minor Cleanup
		ITS#7080
		ITS#7249
		ITS#9934
		ITS#10020
		ITS#10168
		ITS#10226
		ITS#10279
		ITS#10299
		ITS#10302
		ITS#10309
		ITS#10312
		ITS#10320
		ITS#10325
		ITS#10327
		ITS#10328
		ITS#10331
		ITS#10336

OpenLDAP 2.6.9 Release (2024/11/26)
	Fixed libldap TLS connection timeout handling (ITS#8047)
	Fixed libldap GnuTLS incompatible pointer type (ITS#10253)
	Fixed libldap OpenSSL set_ciphersuite error handling (ITS#10223)
	Fixed libldap to check for OpenSSL EVP_Digest* failure (ITS#10224)
	Fixed slapd cn=config disallowed modification of cn=schema (ITS#10256)
	Fixed slapd syncrepl assert during refresh at shutdown (ITS#10232)
	Fixed slapd syncrepl retry state during refreshDone (ITS#10234)
	Fixed slapd-ldap use of multi-precision add for op counters (ITS#10237)
	Fixed slapd-mdb idl intersection (ITS#10233)
	Fixed slapd-wt idl intersection (ITS#10233)
	Fixed slapo-memberof to omit dynamic values (ITS#10230)
	Fixed slapo-nestgroup leak in nestgroup_memberFilter (ITS#10249)
	Fixed slapo-translucent regression with subordinate databases (ITS#10248)
	Fixed slapo-translucent regression when requesting attributes (ITS#10272)
	Fixed slappw-argon2 defaults to be more secure (ITS#9827)
	Minor Cleanup
		ITS#10155
		ITS#10218
		ITS#10219
		ITS#10227
		ITS#10231
		ITS#10235
		ITS#10263
		ITS#10264

OpenLDAP 2.6.8 Release (2024/05/21)
	Fixed libldap exit handling with OpenSSL3 again (ITS#9952)
	Fixed libldap OpenSSL channel binding digest (ITS#10216)
	Fixed slapd handling of large uid/gids peercred auth (ITS#10211)
	Fixed slapd-asyncmeta/meta target structure allocations (ITS#10197)
	Fixed slapd-meta with dynlist (ITS#10164)
	Fixed slapd-meta binds when proxying internal op (ITS#10165)
	Added slapo-nestgroup overlay (ITS#10161)
	Added slapo-memberof 'addcheck' option (ITS#10167)
	Fixed slapo-accesslog startup initialization (ITS#10170)
	Fixed slapo-constraint double free on invalid attr (ITS#10204)
	Fixed slapo-dynlist with abandoned operations (ITS#10044)
	Build
		Fixed build with gcc14.x (ITS#10166)
		Fixed back-perl with clang15 (ITS#10177)
		Fixed to reduce systemd dependencies (ITS#10214)
	Contrib
		Added slapo-alias contrib module (ITS#10104, ITS#10182)
		Fixed slapo-autogroup to work with slapo-dynlist (ITS#10185)
		Fixed smbk5pwd implicit function declaration (ITS#10206)
	Documentation
		Fixed slapo-memberof exattr requirements (ITS#7400)
		Fixed slapo-memberof is no longer deprecated (ITS#7400)
	Minor Cleanup
		ITS#9921
		ITS#10103
		ITS#10171
		ITS#10172
		ITS#10173
		ITS#10179
		ITS#10183
		ITS#10186
		ITS#10188
		ITS#10193
		ITS#10209

OpenLDAP 2.6.7 Release (2024/01/29)
	Added slapo-dynlist option to disable filter support (ITS#10025)
	Fixed liblber missing newline on long msg (ITS#10105)
	Fixed libldap exit handling with OpenSSL3 (ITS#9952)
	Fixed libldap with TLS and multiple ldap URIs (ITS#10101)
	Fixed libldap OpenSSL cipher suite handling (ITS#10094)
	Fixed libldap OpenSSL 3.0 and Diffie-Hellman param files (ITS#10124)
	Fixed libldap timestamps on Windows (ITS#10100)
	Fixed lloadd to work when resolv.conf is missing (ITS#10070)
	Fixed lloadd handling of closing connection (ITS#10083)
	Fixed lloadd tiers to be correctly linked on startup (ITS#10142)
	Fixed slapd to honour disclose in matchedDN handling (ITS#10139)
	Fixed slapd handling of regex testing in ACLs (ITS#10089)
	Fixed slapd sync replication with glued database (ITS#10080)
	Fixed slapd local logging on Windows (ITS#10092)
	Fixed slapd-asyncmeta when remote suffix is empty (ITS#10076)
	Fixed slapo-dynlist so it can't be global (ITS#10091)
	Build
		Fixed lloadd type mismatches (ITS#10074)
		Fixed builds for Windows (ITS#10117)
		Fixed build with clang16 (ITS#10123)
	Documentation
		Fixed slapo-homedir(5) attribute name for olcHomedirArchivePath (ITS#10057)
	Minor Cleanup
		ITS#10059
		ITS#10068
		ITS#10098
		ITS#10109
		ITS#10110
		ITS#10129
		ITS#10130
		ITS#10135
		ITS#10143
		ITS#10144
		ITS#10145
		ITS#10153

OpenLDAP 2.6.6 Release (2023/07/31)
	Fixed slapd cn=config incorrect handling of paused (ITS#10045)
	Fixed slapd-meta to account for MOD ops being optional (ITS#10067)
	Fixed slapd-asyncmeta to account for MOD ops being optional (ITS#10067)


servers/slapd/slappasswd.c
/* $OpenLDAP$ */
/* This work is part of OpenLDAP Software <http://www.openldap.org/>.
 *
 * Copyright 1998-2026 The OpenLDAP Foundation.
 * Portions Copyright 1998-2003 Kurt D. Zeilenga.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted only as authorized by the OpenLDAP
 * Public License.
 *
 * A copy of this license is available in file LICENSE in the
 * top-level directory of the distribution or, alternatively, at
 * <http://www.OpenLDAP.org/license.html>.
 */
/* ACKNOWLEDGEMENTS:
 * This work was initially developed by Kurt Zeilenga for inclusion
 * in OpenLDAP Software.
 */

#include "portable.h"

#include <stdio.h>

#include <ac/stdlib.h>

#include <ac/ctype.h>
#include <ac/signal.h>
#include <ac/socket.h>
#include <ac/string.h>
#include <ac/time.h>
#include <ac/unistd.h>

#include <ldap.h>
#include <lber_pvt.h>
#include <lutil.h>
#include <lutil_sha1.h>

#include "ldap_defaults.h"

#include "slap.h"
#include "slap-config.h"
#include "slapcommon.h"

static char	*modulepath = NULL;
static char	*moduleload = NULL;
static int	moduleargc = 0;
static char	**moduleargv = NULL;

static void
usage(const char *s)
{
	fprintf(stderr,
		"Usage: %s [options]\n"
		"  -c format\tcrypt(3) salt format\n"
		"  -g\t\tgenerate random password\n"
		"  -h hash\tpassword scheme\n"
		"  -n\t\tomit trailing newline\n"
		"  -o <opt>[=val] specify an option with a(n optional) value\n"
		"  \tmodule-path=<pathspec>\n"
		"  \tmodule-load=<filename>\n"
		"  -s secret\tnew password\n"
		"  -u\t\tgenerate RFC2307 values (default)\n"
		"  -v\t\tincrease verbosity\n"
		"  -T file\tread file for new password\n"
		, s );

	exit( EXIT_FAILURE );
}

static int
parse_slappasswdopt( void )
{
	size_t	len = 0;
	char	*p;

	p = strchr( optarg, '=' );
	if ( p != NULL ) {
		len = p - optarg;
		p++;
	}

	if ( strncasecmp( optarg, "module-path", len ) == 0 ) {
		modulepath = p;

	} else if ( strncasecmp( optarg, "module-load", len ) == 0 ) {
		ConfigArgs c = { .line = p };

		if ( config_fp_parse_line( &c ) ) {
			return -1;
		}
		moduleload = c.argv[0];

		moduleargc = c.argc - 1;
		if ( moduleargc ) {
			moduleargv = c.argv+1;
		}

	} else {
		return -1;
	}

	return 0;
}

int
slappasswd( int argc, char *argv[] )
{
	int rc = EXIT_SUCCESS;
#ifdef LUTIL_SHA1_BYTES
	char	*default_scheme = "{SSHA}";
#else
	char	*default_scheme = "{SMD5}";
#endif
	char	*scheme = default_scheme;

	char	*newpw = NULL;
	char	*pwfile = NULL;
	const char *text;
	const char *progname = "slappasswd";

	int		i;
	char		*newline = "\n";
	struct berval passwd = BER_BVNULL;
	struct berval hash = BER_BVNULL;

#ifdef LDAP_DEBUG
	/* tools default to "none", so that at least LDAP_DEBUG_ANY
	 * messages show up; use -d 0 to reset */
	slap_debug = LDAP_DEBUG_NONE;
#endif
	ldap_syslog = 0;

	while( (i = getopt( argc, argv,
		"c:d:gh:no:s:T:vu" )) != EOF )
	{
		switch (i) {
		case 'c':	/* crypt salt format */
			scheme = "{CRYPT}";
			lutil_salt_format( optarg );
			break;

		case 'g':	/* new password (generate) */
			if ( pwfile != NULL ) {
				fprintf( stderr, "Option -g incompatible with -T\n" );
				return EXIT_FAILURE;

			} else if ( newpw != NULL ) {
				fprintf( stderr, "New password already provided\n" );
				return EXIT_FAILURE;

			} else if ( lutil_passwd_generate( &passwd, 8 )) {
				fprintf( stderr, "Password generation failed\n" );
				return EXIT_FAILURE;
			}
			break;

		case 'h':	/* scheme */
			if ( scheme != default_scheme ) {
				fprintf( stderr, "Scheme already provided\n" );
				return EXIT_FAILURE;

			} else {
				scheme = optarg;
			}
			break;

		case 'n':
			newline = "";
			break;

		case 'o':
			if ( parse_slappasswdopt() ) {
				usage ( progname );
			}
			break;

		case 's':	/* new password (secret) */
			if ( pwfile != NULL ) {
				fprintf( stderr, "Option -s incompatible with -T\n" );
				return EXIT_FAILURE;

			} else if ( newpw != NULL ) {
				fprintf( stderr, "New password already provided\n" );
				return EXIT_FAILURE;

			} else {
				char* p;
				newpw = ch_strdup( optarg );

				for( p = optarg; *p != '\0'; p++ ) {
					*p = '\0';
				}
			}
			break;

		case 'T':	/* password file */
			if ( pwfile != NULL ) {
				fprintf( stderr, "Password file already provided\n" );
				return EXIT_FAILURE;

			} else if ( newpw != NULL ) {
				fprintf( stderr, "Option -T incompatible with -s/-g\n" );
				return EXIT_FAILURE;

			}
			pwfile = optarg;
			break;

		case 'u':	/* RFC2307 userPassword */
			break;

		case 'v':	/* verbose */
			verbose++;
			break;

		default:
			usage ( progname );
		}
	}
	slapTool = SLAPPASSWD;

	if( argc - optind != 0 ) {
		usage( progname );
	}

#ifdef SLAPD_MODULES
	if ( module_init() != 0 ) {
		fprintf( stderr, "%s: module_init failed\n", progname );
		return EXIT_FAILURE;
	}

	if ( modulepath && module_path(modulepath) ) {
		rc = EXIT_FAILURE;
		goto destroy;
	}

	if ( moduleload && module_load(moduleload, moduleargc, moduleargv) ) {
		rc = EXIT_FAILURE;
		goto destroy;
	}
#endif

	if( pwfile != NULL ) {
		if( lutil_get_filed_password( pwfile, &passwd )) {
			rc = EXIT_FAILURE;
			goto destroy;
		}
	} else if ( BER_BVISEMPTY( &passwd )) {
		if( newpw == NULL ) {
			/* prompt for new password */
			char *cknewpw;
			newpw = getpassphrase("New password: ");
			if ( newpw == NULL ) { /* Allow EOF to exit. */
				rc = EXIT_FAILURE;
				goto destroy;
			}
			newpw = ch_strdup(newpw);
			cknewpw = getpassphrase("Re-enter new password: ");
			if( cknewpw == NULL || strcmp( newpw, cknewpw )) {
				fprintf( stderr,
				    "Password values do not match\n" );
				rc = EXIT_FAILURE;
				goto destroy;
			}
		}

		passwd.bv_val = newpw;
		passwd.bv_len = strlen(passwd.bv_val);
	} else {
		hash = passwd;
		goto print_pw;
	}

	lutil_passwd_hash( &passwd, scheme, &hash, &text );
	if ( BER_BVISNULL( &hash ) ) {
		fprintf( stderr,
			"Password generation failed for scheme %s: %s\n",
			scheme, text ? text : "" );
		rc = EXIT_FAILURE;
		goto destroy;
	}

	if( lutil_passwd( &hash, &passwd, NULL, &text ) ) {
		fprintf( stderr, "Password verification failed. %s\n",
			text ? text : "" );
		rc = EXIT_FAILURE;
		goto destroy;
	}

print_pw:;
	printf( "%s%s" , hash.bv_val, newline );

destroy:;
#ifdef SLAPD_MODULES
	module_kill();
#endif
	if ( !BER_BVISNULL( &hash ) ) {
		ber_memfree( hash.bv_val );
	}
	if ( passwd.bv_val != hash.bv_val && !BER_BVISNULL( &passwd ) ) {
		ber_memfree( passwd.bv_val );
	}

	return rc;
}
